From 35218cb7a44fd45c4ecefbeb93b8a0723d53cd85 Mon Sep 17 00:00:00 2001
From: Ben Greear <greearb@candelatech.com>
Date: Mon, 19 Jul 2021 10:40:14 -0700
Subject: [PATCH] mt76: mt7915: txfree status to show txcount instead of
 latency

Latency is not obviously that useful, but txcount can let us deduce
retries, which may be more interesting.

Signed-off-by: Ben Greear <greearb@candelatech.com>
---
 mt7915/init.c | 3 +++
 mt7915/mac.h  | 8 +++++---
 mt7915/regs.h | 4 ++++
 3 files changed, 12 insertions(+), 3 deletions(-)

Index: mt76-2021-10-20-b649678c/mt7915/init.c
===================================================================
--- mt76-2021-10-20-b649678c.orig/mt7915/init.c
+++ mt76-2021-10-20-b649678c/mt7915/init.c
@@ -402,6 +402,9 @@ static void mt7915_mac_init(struct mt791
 	/* enable hardware de-agg */
 	mt76_set(dev, MT_MDP_DCR0, MT_MDP_DCR0_DAMSDU_EN);
 
+	/* disable Tx latency report to enable Tx count in txfree path */
+	mt76_clear(dev, MT_PLE_HOST_RPT0, MT_PLE_HOST_RPT0_TX_LATENCY);
+
 	for (i = 0; i < MT7915_WTBL_SIZE; i++)
 		mt7915_mac_wtbl_update(dev, i,
 				       MT_WTBL_UPDATE_ADM_COUNT_CLEAR);
Index: mt76-2021-10-20-b649678c/mt7915/mac.h
===================================================================
--- mt76-2021-10-20-b649678c.orig/mt7915/mac.h
+++ mt76-2021-10-20-b649678c/mt7915/mac.h
@@ -305,13 +305,15 @@ struct mt7915_tx_free {
 
 #define MT_TX_FREE_MSDU_CNT		GENMASK(9, 0)
 #define MT_TX_FREE_WLAN_ID		GENMASK(23, 14)
+/* when configured for txfree latency mode.  See MT_PLE_HOST_RPT0_TX_LATENCY
+ * Not enabled by default now.
+ */
 #define MT_TX_FREE_LATENCY		GENMASK(12, 0)
-/* 0: success, others: dropped */
+/* when configured for txcount mode.  See MT_PLE_HOST_RPT0_TX_LATENCY. */
+#define MT_TX_FREE_TXCNT		GENMASK(12, 0)
 #define MT_TX_FREE_STATUS		GENMASK(14, 13)
 #define MT_TX_FREE_MSDU_ID		GENMASK(30, 16)
 #define MT_TX_FREE_PAIR			BIT(31)
-/* will support this field in further revision */
-#define MT_TX_FREE_RATE			GENMASK(13, 0)
 
 #define MT_TXS0_FIXED_RATE		BIT(31)
 #define MT_TXS0_BW			GENMASK(30, 29)
Index: mt76-2021-10-20-b649678c/mt7915/regs.h
===================================================================
--- mt76-2021-10-20-b649678c.orig/mt7915/regs.h
+++ mt76-2021-10-20-b649678c/mt7915/regs.h
@@ -22,6 +22,10 @@
 #define MT_PLE_BASE			0x8000
 #define MT_PLE(ofs)			(MT_PLE_BASE + (ofs))
 
+/* Modify whether txfree struct returns latency or txcount. */
+#define MT_PLE_HOST_RPT0				MT_PLE(0x030)
+#define MT_PLE_HOST_RPT0_TX_LATENCY		BIT(3)
+
 #define MT_FL_Q_EMPTY			0x0b0
 #define MT_FL_Q0_CTRL			0x1b0
 #define MT_FL_Q2_CTRL			0x1b8
